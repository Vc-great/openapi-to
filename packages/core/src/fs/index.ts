export { clean } from './clean.ts'
export { pathExistsSync } from './pathExistsSync.ts'
export { read, readSync } from './read.ts'
export { write } from './write.ts'
