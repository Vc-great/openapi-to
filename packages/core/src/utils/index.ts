export * from "./cache.ts";
export * from "./cache.ts";
export * from "./clean.ts";
export * from "./logger.ts";
export * from "./pathExistsSync.ts";
export * from "./promise.ts";
export * from "./Queue.ts";
export * from "./Queue.ts";
export * from "./randomColour.ts";
export * from "./read.ts";
export * from "./renderTemplate.ts";
export * from "./throttle.ts";
export * from "./timeout.ts";
export * from "./transformers/index.ts";
export * from "./uniqueName.ts";
export * from "./URLPath.ts";
export * from "./write.ts";
