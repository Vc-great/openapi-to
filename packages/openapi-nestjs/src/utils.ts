import stringifyObject from "./stringifyObject.ts";

export function formatterType(type: string): string {
  if (type === "integer") {
    return "number";
  }
  return type;
}

export function objectToStringify(object: unknown): string {
  return stringifyObject(object, {
    filter(obj: { [k: string]: unknown }, key: string) {
      return obj[key] !== undefined;
    },
  });
}
