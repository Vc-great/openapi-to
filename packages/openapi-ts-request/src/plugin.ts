import {AST,OpenAPI,createPlugin} from '@openapi-to/core'
import {RequestGenerator} from "./RequestGenerator.ts";
import Oas from 'oas';
import type {PluginConfig} from './types.ts'

export const definePlugin = createPlugin<PluginConfig>((pluginConfig)=>({openapiDocument,openapiToSingleConfig}) => {
  const ast = new AST()
  const oas = new Oas({...openapiDocument})
  const openapi = new OpenAPI({}, oas);
  return {
    name:"openapi-ts-request",
   buildStart(){
     const requestGenerator = new RequestGenerator({
       oas,
       ast,
       openapi,
       pluginConfig,
       openapiToSingleConfig,
     });
     requestGenerator.build()
    },
    writeFile(){
      ast.saveSync()
    },
    buildEnd(){
      //log
    }
  }

})
