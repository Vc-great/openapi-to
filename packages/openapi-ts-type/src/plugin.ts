import { AST, createPlugin, OpenAPI } from "@openapi-to/core";

import Oas from "oas";

import { TypeGenerator } from "./TypeGenerator.ts";

import type { PluginConfig } from "./types.ts";

export const definePlugin = createPlugin(
  (pluginConfig?: PluginConfig) =>
    ({ openapiDocument, openapiToSingleConfig }) => {
      const ast = new AST();
      const oas = new Oas({ ...openapiDocument });
      const openapi = new OpenAPI({}, oas);
      return {
        name: "openapi-ts-type",
        buildStart(context) {
          const requestGenerator = new TypeGenerator({
            oas,
            ast,
            openapi,
            pluginConfig,
            openapiToSingleConfig,
          });
          requestGenerator.build(context);
        },
        writeFile() {
          return ast.saveSync();
        },
        buildEnd() {},
      };
    },
);
