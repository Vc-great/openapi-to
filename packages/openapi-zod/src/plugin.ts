import { AST, createPlugin, OpenAPI } from "@openapi-to/core";

import Oas from "oas";

import { ZodGenerator } from "./ZodGenerator.ts";

import type { PluginConfig } from "./types.ts";

export const definePlugin = createPlugin(
  (pluginConfig?: PluginConfig) =>
    ({ openapiDocument, openapiToSingleConfig }) => {
      const ast = new AST();
      const oas = new Oas({ ...openapiDocument });
      const openapi = new OpenAPI({}, oas);
      return {
        name: "openapi-zod",
        buildStart() {
          const typeGenerator = new ZodGenerator({
            oas,
            ast,
            openapi,
            pluginConfig,
            openapiToSingleConfig,
          });
          typeGenerator.build();
        },
        writeFile() {
          return ast.saveSync();
        },
        buildEnd() {},
      };
    },
);
