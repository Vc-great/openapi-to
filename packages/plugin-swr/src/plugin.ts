import { createPlugin, pluginEnum } from '@openapi-to/core'

import path from 'node:path'
import { kebabCase, upperFirst } from 'lodash-es'
import { Project, StructureKind } from 'ts-morph'
import { buildImports } from './builders/buildImports.ts'
import { buildMethodBody } from './builders/buildMethodBody.ts'
import { buildMethodParameters } from './builders/buildMethodParameters.ts'
import { buildQueryKey, buildQueryKeyType } from './builders/buildQueryKey.ts'
import { jsDocTemplateFromMethod } from './templates/jsDocTemplateFromMethod.ts'
import type { PluginConfig } from './types.ts'

export const definePlugin = createPlugin<PluginConfig>((_pluginConfig) => {
  const project = new Project()
  const pluginConfig: PluginConfig = {
    ..._pluginConfig,
    responseConfigTypeImportDeclaration: {
      namedImports: _pluginConfig?.responseConfigTypeImportDeclaration?.namedImports ?? ['AxiosResponse'],
      moduleSpecifier: _pluginConfig?.responseConfigTypeImportDeclaration?.moduleSpecifier ?? 'axios',
    },
    responseErrorTypeImportDeclaration: {
      namedImports: _pluginConfig?.responseErrorTypeImportDeclaration?.namedImports ?? ['AxiosError'],
      moduleSpecifier: _pluginConfig?.responseErrorTypeImportDeclaration?.moduleSpecifier ?? 'axios',
    },
    importWithExtension: _pluginConfig?.importWithExtension ?? true,
  }
  return {
    name: pluginEnum.SWR,
    dependencies: [pluginEnum.TsType, pluginEnum.Request],
    hooks: {
      buildStart: async (ctx) => {
        // 可注入日志、校验 pluginConfig
        // ctx.logger.info('Request 插件启动', pluginConfig)
      },
      operation: async (operation, ctx) => {
        const hookName = `use${upperFirst(operation.accessor.operationName)}`
        const folderName = operation.accessor.getFirstTagName ?? ''

        const filePath = path.join(ctx.openapiToSingleConfig.output.dir, folderName, `${kebabCase(hookName)}.swr.ts`)
        const operationSourceFile = project.createSourceFile(filePath, '', { overwrite: true })

        operationSourceFile.addStatements(buildImports(filePath, operation, pluginConfig))

        //key
        operationSourceFile.addStatements([buildQueryKey(operation, pluginConfig), buildQueryKeyType(operation)])

        operationSourceFile.addFunction({
          kind: StructureKind.Function,
          isAsync: false,
          name: hookName,
          parameters: buildMethodParameters(operation, pluginConfig),
          returnType: undefined,
          isExported: true,
          docs: jsDocTemplateFromMethod(operation),
          statements: buildMethodBody(operation, pluginConfig),
        })

        ctx.setSourceFiles([pluginEnum.SWR, operation.accessor.operationName], operationSourceFile)
      },
      tagEnd: async (tagData, ctx) => {},
      buildEnd() {},
    },
  }
})
