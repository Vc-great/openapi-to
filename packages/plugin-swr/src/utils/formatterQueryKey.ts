import type { OperationWrapper } from '@openapi-to/core'
import { upperFirst } from 'lodash-es'
import { OpenAPIV3 } from 'openapi-types'

export const formatterQueryKeyTypeName = (operation: OperationWrapper) =>
  `${upperFirst(operation.accessor.operationName)}${operation.method === OpenAPIV3.HttpMethods.GET ? 'Query' : 'Mutation'}Key`

export const formatterQueryKeyName = (operation: OperationWrapper) =>
  `${operation.accessor.operationName}${operation.method === OpenAPIV3.HttpMethods.GET ? 'Query' : 'Mutation'}Key`

//
