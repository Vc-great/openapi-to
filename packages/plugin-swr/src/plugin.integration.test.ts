//@ts-nocheck
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { PluginManager, pluginEnum } from '@openapi-to/core'
import { checkFolderHasFiles } from '@openapi-to/core/utils'

import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import mockOpenAPI from '../mock/petstore.json'
import { definePlugin } from './plugin'

// 导入TsType插件
import { definePlugin as defineTsTypePlugin } from '@openapi-to/plugin-ts-type'
import { definePlugin as defineTsRequestPlugin } from '@openapi-to/plugin-ts-request'

// 获取当前文件的目录
const __dirname = path.dirname(fileURLToPath(import.meta.url))
const TEST_OUTPUT_DIR = path.resolve(__dirname, '../test-output')

describe('swr Plugin Integration', () => {
  beforeEach(() => {
    // 清理并重新创建测试输出目录
    if (fs.existsSync(TEST_OUTPUT_DIR)) {
      fs.rmSync(TEST_OUTPUT_DIR, { recursive: true, force: true })
    }
    fs.mkdirSync(TEST_OUTPUT_DIR, { recursive: true })

    console.log(`测试目录已准备: ${TEST_OUTPUT_DIR}`)
  })

  afterEach(() => {
    // 确保测试目录存在
    if (!fs.existsSync(TEST_OUTPUT_DIR)) {
      fs.mkdirSync(TEST_OUTPUT_DIR, { recursive: true })
    }
  })

  it('应该生成默认的服务类文件', async () => {
    const pluginManager = new PluginManager(
      {
        name: 'swr',
        root: '',
        plugins: [defineTsTypePlugin(), defineTsRequestPlugin(),definePlugin()],
        input: {
          path: '',
        },
        output: {
          dir: TEST_OUTPUT_DIR,
        },
      },
      // @ts-ignore
      mockOpenAPI,
    )

    await pluginManager.run()

    // 检查文件是否生成
    const hasFiles = checkFolderHasFiles(TEST_OUTPUT_DIR)
    expect(hasFiles).toBe(true)

    // 修正为正确的文件路径 - 检查pet目录
    const petDir = path.join(TEST_OUTPUT_DIR, 'pet')
    expect(fs.existsSync(petDir)).toBe(true)

    // 查找pet目录下的所有service文件
    const swrFiles = fs.readdirSync(petDir).filter((file) => file.endsWith('swr.ts'))
    expect(swrFiles.length).toBeGreaterThan(0)

    // 选取第一个服务文件进行内容检查
    const petSwrPath = path.join(petDir, 'use-add-pet.swr.ts')
    const fileContent = fs.readFileSync(petSwrPath, 'utf-8')

    expect(fileContent).toContain('useAddPet')

    expect(fileContent).toMatchSnapshot()
  })
})
