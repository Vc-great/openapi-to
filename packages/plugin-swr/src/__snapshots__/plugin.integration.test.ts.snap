// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`swr Plugin Integration > 应该生成默认的服务类文件 1`] = `
"import useSWRMutation from "swr/mutation";
import type { SWRMutationConfiguration } from "swr/mutation";
import type { AddPetMutationRequest, AddPetMutationResponse, AddPetResponseError } from "./add-pet.types.ts";
import { addPetService } from "./add-pet.service.ts";
export const addPetMutationKey = () => [{ url: '/pet', method: 'post' }] as const;
export type AddPetMutationKey = ReturnType<typeof addPetMutationKey>;

/**
 * @summary Add a new pet to the store
 */
export function useAddPet(options?: {
  mutation?: SWRMutationConfiguration<AddPetMutationResponse, AddPetResponseError, AddPetMutationKey | null, AddPetMutationRequest>;
  shouldFetch?: boolean;
}) {

  const { mutation: mutationOptions, shouldFetch = true } = options ?? {};
  const mutationKey = addPetMutationKey();

  return useSWRMutation<
    AddPetMutationResponse,
    AddPetResponseError,
    AddPetMutationKey | null,
    AddPetMutationRequest
  >(
    shouldFetch ? mutationKey : null,
    async (_url, { arg: data }) => {
      return addPetService(data);
    },
    mutationOptions
  );
}
"
`;
